src/nvtxw.rs

```diff
@@ -0,0 +1,419 @@
+            assert!(row.len() == meta_row.len());
+
+            for (item, meta_item) in zip(row, meta_row) {
+                let time_start = item.interval.start;
+                let time_stop = item.interval.stop;
+                let color = item.color;
+                // let time_start = meta_item.original_interval.start;
+                // let time_stop = meta_item.original_interval.stop;
+                let title = meta_item.title.clone();
+
+                let c_name = CString::new(title).expect("CString::new failed");
+                let events = [legion_nvtxw_event {
+                    time_start: (time_start.0 as u64)
+                        .checked_add(zero_time.try_into().unwrap())
+                        .expect("time_start overflowed"),
+                    time_stop: (time_stop.0 as u64)
+                        .checked_add(zero_time.try_into().unwrap())
+                        .expect("time_stop overflowed"),
+                    name: c_name.as_ptr(),
+                    color: ((color.r() as u32) << 16)
+                        | ((color.g() as u32) << 8)
+                        | (color.b() as u32)
+                        | (0xFF << 24),
+                }];
+
+                let stream = streams[&tile.entry_id];
+
+                let payloads = [
+                    nvtxw::PayloadData {
+                        schemaId: LEGION_NVTXW_PAYLOAD_NAME_SCHEMA_ID,
+                        size: usize::MAX,
+                        payload: c_name.as_ptr() as *const c_void,
+                    },
+                    nvtxw::PayloadData {
+                        schemaId: LEGION_NVTXW_PAYLOAD_SCHEMA_ID,
+                        size: size_of::<legion_nvtxw_event>(),
+                        payload: events.as_ptr() as *const c_void,
+                    },
+                ];
+
+                nvtxw::event_write(interface, stream, &payloads).expect("Failed to write event");
+            }
+        }
+    }
+
+    fn process_events(
+        data_source: &mut CountingDeferredDataSource<T>,
+        interface: nvtxw::InterfaceHandle,
+        streams: &BTreeMap<EntryID, nvtxw::StreamHandle>,
+        zero_time: i64,
+        unmatched_tiles: &mut UnmatchedTileHold,
+        num_requests: u64,
+    ) {
+        while data_source.outstanding_requests() > num_requests {
+            // let summary_tiles = data_source.get_summary_tiles();
+            let slot_tiles = data_source.get_slot_tiles();
+            let slot_meta_tiles = data_source.get_slot_meta_tiles();
+
+            for tile in slot_tiles {
+                match unmatched_tiles.entry(tile.entry_id.clone()) {
+                    Entry::Occupied(mut entry) => {
+                        let (ut, _um) = entry.get_mut();
+                        *ut = Some(tile);
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert((Some(tile), None));
+                    }
+                }
```

Comment:
I think you could shorten this to:

```rust
unmatched_tiles.entry(tile.entry_id.clone()).or_insert((None, None)).0 = Some(tile);
```

That is, the `or_insert` ensures we get a default value if it's missing, but otherwise we're just going to overwrite it anyway, so no need to branch.

And similarly for the meta case but make sure to access `.1` (i.e., the second field in the tuple).

Reply:
Done. I'm really glad there is a simpler version of this.